### PR TITLE
fixes #11702 - backend refactor for content units

### PR DESCRIPTION
**Pull request by jlsherrill**
This refactors the backend content units to isolate backend calls into a class instead of a module that is included in the matching AR model. 

This allows for better isolation and makes it more obvious when a piece of data is from a backend service.  I would like to do this with consumers as part of host unification and wanted to get feedback.

### CODE REVIEW

**Comment by jlsherrill** [comment]
Lots of tests will break, but curious about feedback.  Will also create an issue if comments are positive.

**Comment by daviddavis** [comment]
:+1:

**Comment by ehelms** [comment]
Generally speaking, I think its a solid change. What about moving to `app/backend/pulp` or `app/backends/pulp` or rooting at `app/services/pulp`? This would follow similar conventions, and allow the `app/models` directory to move toward being just the normal Rails/active record model bits. And since other layers than the model layer, such as the actions layer, will use the backend layer directly this makes sense to move out to me.

**Comment by jlsherrill** [comment]
@ehelms i do like that idea :+1:  

will update

File: app/models/katello/concerns/pulp_content_unit.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+require 'set'
+
+module Katello
+  module Concerns::PulpContentUnit
```
Comment (ehelms) [inline app/models/katello/concerns/pulp_content_unit.rb]: Should this be moved into app/services as well? Seems very Pulp layer specific and only being used by them.

File: app/views/katello/api/v2/packages/show.json.rabl [hunk 2]
```diff
@@ -1,11 +1,4 @@
...
 extends "katello/api/v2/packages/base"
-
-attributes :description
-attributes :license, :buildhost, :vendor, :relativepath
-attributes :children, :checksumtype, :size, :url, :build_time, :summary, :group, :requires, :provides, :files
-
-node :human_readable_size do |package|
-  number_to_human_size(package.size)
-end
+extends "katello/api/v2/packages/backend", :object => Katello::Pulp::Rpm.new(@resource.uuid)
```
Comment (ehelms) [inline app/views/katello/api/v2/packages/show.json.rabl]: Did you consider potentially doing at the top of, say the RPM backend rabl:

```
object Katello::Pulp::Rpm.new(@resource.uuid)
``` 

I know the current method keeps the backend rabl agnostic, I'm just not sure I ever see it being used without doing the above transformation? Just a thought, as I am 50/50 on each idea.

Comment (jlsherrill) [inline app/views/katello/api/v2/packages/show.json.rabl]: I think its cleaner the way i did it, but i don't really have any strong argument either way.

File: app/services/katello/pulp/pulp_content_unit.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
+require 'set'
+
+module Katello
+  module Pulp
+    module PulpContentUnit
```
Comment (ehelms) [inline app/services/katello/pulp/pulp_content_unit.rb]: I was curious your thinking on making this an composable object instead of a base class the content units inherit from. What prompted me to ask the question was the fact that this module had an initialize method.

File: test/services/katello/pulp/rpm_test.rb [hunk 4]
```diff
@@ -2,14 +2,14 @@
 require 'support/pulp/repository_support'
 
 module Katello
-  class GluePulpRpmTestBase < ActiveSupport::TestCase
+  class ServicesPulpRpmTestBase < ActiveSupport::TestCase
     include RepositorySupport
 
     @@package_id = nil
```
Comment (ehelms) [inline test/services/katello/pulp/rpm_test.rb]: Is this used anywhere?

File: test/services/katello/pulp/rpm_test.rb [hunk 5]
```diff
@@ -30,25 +30,25 @@ def teardown
     end
   end
 
-  class GluePulpRpmTest < GluePulpRpmTestBase
+  class ServicesGluePulpRpmTest < ServicesPulpRpmTestBase
```
Comment (ehelms) [inline test/services/katello/pulp/rpm_test.rb]: Remove glue

File: test/services/katello/pulp/rpm_test.rb [hunk 6]
```diff
@@ -2,14 +2,14 @@
 require 'support/pulp/repository_support'
 
 module Katello
-  class GluePulpRpmTestBase < ActiveSupport::TestCase
+  class ServicesPulpRpmTestBase < ActiveSupport::TestCase
```
Comment (ehelms) [inline test/services/katello/pulp/rpm_test.rb]: What about wrapping this in module declaration for Services and Pulp vs this long test name that encompasses various layers?

**Comment by ehelms** [comment]
This may be left field, but thinking out loud. With this you introduce the notion of a backend rabl view which I think is a great way to encompass. However, what we don't have is a way to vet that the show or base templates do NOT include backend data and that the backend templates only include backend (lazy accessor) data. What are your thoughts on building some tests that use the direct rabl render (https://paste.soc.redhat.com/6881/) to assert that what is or is not included matches up against attributes defined by the lazy accessor methods?

**Comment by daviddavis** [comment]
@ehelms please use a publicly accessible sites upstream.

**Comment by ehelms** [comment]
@daviddavis I apparently didn't have the link I thought I had in my clipboard, as even if you could access it, it wouldn't have made any sense. I updated the comment with an edit note.

**Comment by daviddavis** [comment]
I'm not opposed to rabl testing but I would be weary of trivial tests that simply check that attributes are present. I think testing declarative logic is largely pointless unless there is sufficient complexity involved.

**Comment by ehelms** [comment]
I think my point was missed. I'm asking about using tests to ensure that
the back end template contains only backend attributes (I.e. those that are
lazy loaded) and that the base and show templates do not as a way to
prevent unnecessary / unknown calls to backend systems for data. Since that
has led to some performance issues.

**Comment by ehelms** [comment]
After some discussion, @daviddavis asked me for an example of what I meant so I am re-posting here. This would test that the backend template only contains backend attributes and that the base template does NOT contain backend attributes. We'd likely want a similar test for 'show'.

```
@errata = Errata.first
rendered = Rabl::Renderer.new('errata/backend', @errata, :view_path => 'app/views/katello', :format => 'hash').render

assert_empty rendered.keys - Pulp::PackageGroup::BACKEND_ATTRIBUTES

rendered = Rabl::Renderer.new('errata/base', @errata, :view_path => 'app/views/katello', :format => 'hash').render

assert_empty rendered.keys & Pulp::PackageGroup::BACKEND_ATTRIBUTES
```

**Comment by jlsherrill** [comment]
@ehelms so i agree with testing these rabl templates, i think i disagree with what is being tested.  I'd rather write a test that just verifies the rabl can be rendered without error with no stubs or mocks in place.  

I would rather not maintain a list of backend attributes and compare the rendered rabl to that.  That seems like a silly test.

**Comment by ehelms** [comment]
That would suffice as well, however, I don't see how its a silly test given your reasoning as you are already maintaining a list of backend attributes -- https://github.com/Katello/katello/pull/5461/files#diff-2812c1b5663e29b173cad58ed2760ca0R9

**Comment by jlsherrill** [comment]
@ehelms yes, but the lazy accessors do not necessarily map 100% to what is in the backend rabl.  For example packages/backend includes a human_readable_size.  Its possible that attributes may be massaged or transformed.

**Comment by ehelms** [comment]
That's fair - could you show an example of your test idea?

**Comment by jlsherrill** [comment]
@ehelms @daviddavis 

thoughts on this:
```
require 'katello_test_helper'

module Katello
  class RpmViewTest < ActiveSupport::TestCase

    def setup
      @rpm = katello_rpms(:two)
    end

    def test_base
      assert_service_not_used(Pulp::Rpm) do
        render_rabl('katello/api/v2/packages/base.json', @rpm)
      end
    end

    def test_show
      assert_service_used(Pulp::Rpm) do
        render_rabl('katello/api/v2/packages/show.json', @rpm)
      end
    end

    def render_rabl(filepath, resource)
      Rabl::Renderer.new(filepath, resource, :view_path => "#{Katello::Engine.root}/app/views/",
                         :format => 'hash', :locals => {:resource => resource}).render
    end

    def assert_service_not_used(service_class)
      service_class.any_instance.expects(:backend_data).never
      yield
    end

    def assert_service_used(service_class)
      service_class.any_instance.expects(:backend_data).returns({})
      yield
    end
  end
end
```

**Comment by ehelms** [comment]
I think that test looks alright, it doesn't prevent the case where backend template could be used where it shouldn't but perhaps that case is an edge case. I could see one advantage to putting the Pulp::Rpm declaration into the backend template is that anything that did try to use the backend template inappropriately would be caught in tests by the application attempting to contact Pulp.

**Comment by jlsherrill** [comment]
@ehelms why would the same not be true of these tests?

**Comment by ehelms** [comment]
OK, I think the tests and method you propose works for me. Add those and then I think this will be good to go.

File: app/models/katello/concerns/pulp_database_unit.rb [hunk 7]
```diff
@@ -16,8 +12,8 @@ def backend_data
     end
 
     module ClassMethods
-      def unit_handler
-        Katello.pulp_server.extensions.send(self.name.demodulize.underscore)
+      def content_unit_class
+        "::Katello::Backend::#{self.name.demodulize}"
```
Comment (ehelms) [inline app/models/katello/concerns/pulp_database_unit.rb]: This was renamed I think?

File: app/services/katello/pulp/pulp_content_unit.rb (hunk 3 above)
Comment (jlsherrill) [inline app/services/katello/pulp/pulp_content_unit.rb]: Honestly i think it was just done that way as that is how pulp_database_unit was done.  I can change it if preferred.  It probably does make more sense

Comment (ehelms) [inline app/services/katello/pulp/pulp_content_unit.rb]: :+1:

On Thu, Oct 1, 2015 at 12:34 PM, Justin Sherrill <notifications@github.com>
wrote:

> In app/services/katello/pulp/pulp_content_unit.rb
> <https://github.com/Katello/katello/pull/5461#discussion_r40935252>:
>
> > @@ -0,0 +1,61 @@
> > +require 'set'
> > +
> > +module Katello
> > +  module Pulp
> > +    module PulpContentUnit
>
> Honestly i think it was just done that way as that is how
> pulp_database_unit was done. I can change it if preferred. It probably does
> make more sense
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Katello/katello/pull/5461/files#r40935252>.
>



-- 
Eric D. Helms
Red Hat Engineering
Ph.D. Student - North Carolina State University

Comment (jlsherrill) [inline app/services/katello/pulp/pulp_content_unit.rb]: updated

File: app/services/katello/pulp/pulp_content_unit.rb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+      def self.fetch(offset, page_size, uuids = nil)
+        fields = ::PULP_INDEXED_FIELDS if self.constants.include?(:PULP_INDEXED_FIELDS)
+        criteria = {:limit => page_size, :skip => offset}
+        criteria[:fields] = fields if fields
+        criteria[:filters] = {'_id' => {'$in' => uuids}} if uuids
+        Katello.pulp_server.resources.unit.search(self::CONTENT_TYPE, criteria, :include_repos => true)
+      end
+
+      def initialize(uuid)
```
Comment (ehelms) [inline app/services/katello/pulp/pulp_content_unit.rb]: Nitpick: Move to the top since its the entry point of the class?

File: app/services/katello/pulp/pulp_content_unit.rb [hunk 9]
```diff
@@ -0,0 +1,55 @@
+require 'set'
+
+module Katello
+  module Pulp
+    class PulpContentUnit
+      #  Any class that includes this module should define:
```
Comment (ehelms) [inline app/services/katello/pulp/pulp_content_unit.rb]: This says they should define whats below, but 2 of the 3 are listed as optional which seems conflicting to me as a developer trying to add a new type.

Comment (jlsherrill) [inline app/services/katello/pulp/pulp_content_unit.rb]: seemed kinda self-explanatory to me, but i guess previously there was more than one non-require method/constant.  add more verbose text

**Comment by ehelms** [comment]
ACK !
